<a href="https://colab.research.google.com/github/fmorales10/Portafolio/blob/main/1_Movilidad.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px
import datetime as dt
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [ ]:
from google.colab import files 
uploaded = files.upload()

Saving afluencia-preliminar-en-transporte-publico.xlsx-afluencia_diaria.csv to afluencia-preliminar-en-transporte-publico.xlsx-afluencia_diaria (1).csv


In [ ]:
df= pd.read_csv("afluencia-preliminar-en-transporte-publico.xlsx-afluencia_diaria.csv")

In [ ]:
df

id  organismo linea_servicio        dia       fecha  \
0          1    Ecobici            NaN    Domingo  01/03/2020   
1          2    Ecobici            NaN      Lunes  02/03/2020   
2          3    Ecobici            NaN     Martes  03/03/2020   
3          4    Ecobici            NaN  Miércoles  04/03/2020   
4          5    Ecobici            NaN     Jueves  05/03/2020   
...      ...        ...            ...        ...         ...   
18709  18710  Suburbano             L1      Lunes  28/06/2021   
18710  18711  Suburbano             L1     Martes  29/06/2021   
18711  18712  Suburbano             L1  Miércoles  30/06/2021   
18712  18713  Suburbano             L1     Jueves  01/07/2021   
18713  18714  Suburbano             L1    Viernes  02/07/2021   

      afluencia_tarjeta afluencia_boleto afluencia_total_preliminar  
0                   NaN              NaN                     11,238  
1                   NaN              NaN                     29,475  
2                   NaN              NaN                     31,855  
3                   NaN              NaN                     31,477  
4                   NaN              NaN                     31,493  
...                 ...              ...                        ...  
18709               NaN              NaN                        NaN  
18710               NaN              NaN                        NaN  
18711               NaN              NaN                        NaN  
18712               NaN              NaN                        NaN  
18713               NaN              NaN                        NaN  

[18714 rows x 8 columns]

Ahora hacemos un conteo de los valores nulos de cada columna.

In [ ]:
df.isnull().sum()

id                                0
organismo                         0
linea_servicio                  489
dia                               0
fecha                             0
afluencia_tarjeta             16027
afluencia_boleto              15116
afluencia_total_preliminar      202
dtype: int64

Podemos notar que la columna con más valores nulos es "afluencia_tarjeta" con 16027 valores nulos.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18714 entries, 0 to 18713
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          18714 non-null  int64 
 1   organismo                   18714 non-null  object
 2   linea_servicio              18225 non-null  object
 3   dia                         18714 non-null  object
 4   fecha                       18714 non-null  object
 5   afluencia_tarjeta           2687 non-null   object
 6   afluencia_boleto            3598 non-null   object
 7   afluencia_total_preliminar  18512 non-null  object
dtypes: int64(1), object(7)
memory usage: 1.1+ MB


In [ ]:
df['fecha']=pd.to_datetime(df['fecha'])
df['afluencia_tarjeta']=df['afluencia_tarjeta'].str.replace(",","")
df['afluencia_boleto']=df["afluencia_boleto"].str.replace(",","")
df['afluencia_total_preliminar']=df['afluencia_total_preliminar'].str.replace(',','')
df['afluencia_boleto']=pd.to_numeric(df['afluencia_boleto'])
df['afluencia_tarjeta']=pd.to_numeric(df['afluencia_tarjeta'])
df['afluencia_total_preliminar']=pd.to_numeric(df['afluencia_total_preliminar'])

No todos los tipos de datos son correctos.

In [ ]:
df.groupby(["organismo"])["afluencia_total_preliminar"].sum().sort_values(ascending = False)

organismo
STC                1.022922e+09
Metrobús           3.019027e+08
RTP                9.093083e+07
STE-Trolebús       4.424603e+07
Suburbano          3.260932e+07
STE-Tren Ligero    1.230674e+07
Ecobici            4.672818e+06
STE-Cablebús       3.850480e+05
Name: afluencia_total_preliminar, dtype: float64

El organismo más utilizado es "STC"

In [ ]:
df.groupby(["linea_servicio"])["afluencia_total_preliminar"].sum().sort_values(ascending = False)

linea_servicio
L1                                                   282218802.0
L2                                                   181470897.0
L3                                                   168816310.0
LB                                                   101715310.0
L5                                                    94596366.0
L8                                                    90264901.0
L7                                                    89482578.0
LA                                                    83532311.0
L12                                                   78835682.0
L9                                                    78760298.0
L6                                                    71889214.0
Ordinario                                             45580256.0
L4                                                    35249988.0
Expreso                                               24647012.0
L1 Eje Central                                        22888645.0
Xochimilco

La linea de servicios mas utilizada es la L1

¿Qué día de la semana tiene más uso el transporte público?

In [ ]:

df['dia']=df["dia"].str.replace("Míercoles","Miércoles")

In [ ]:
df.groupby(["dia"])["afluencia_total_preliminar"].sum().sort_values(ascending = False)

dia
Miércoles    240544976.0
Martes       239477843.0
Jueves       238240944.0
Viernes      237182340.0
Lunes        228317986.0
Sábado       197344970.0
Domingo      128866694.0
Name: afluencia_total_preliminar, dtype: float64

El miércoles es el día con mayor movimiento

Por más de un millón de registros ante el segundo día con más afluencia.

In [ ]:
dfx = pd.DataFrame(((df.isnull().sum() / len(df))*100))
dfx=dfx.reset_index()



In [ ]:
dfx.columns = ['columna','porcentaje']

In [ ]:
dfx.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   columna     8 non-null      object 
 1   porcentaje  8 non-null      float64
dtypes: float64(1), object(1)
memory usage: 256.0+ bytes


In [ ]:
dfx

columna  porcentaje
0                          id    0.000000
1                   organismo    0.000000
2              linea_servicio    2.613017
3                         dia    0.000000
4                       fecha    0.000000
5           afluencia_tarjeta   85.641766
6            afluencia_boleto   80.773752
7  afluencia_total_preliminar    1.079406

In [ ]:
fig = px.bar(dfx, x='columna', y='porcentaje')

In [ ]:
fig

Instrucciones
Responde a las siguientes preguntas: 
Notarás que las columnas de boletos tienen una gran cantidad de valores nulos. ¿Por qué? ¿Se pueden eliminar todos los registros que contengan datos nulos? ¿Por qué sí o por qué no? Contesta


Puede ser porque no hubo un registro suficiente de la cantidad de los boletos, y puede ser que algunos metodos de transporte no se entreguen boletos, puede ser porque se use un sistema de tarjeta etc, se pueden eliminar pero no deberiamos pues hariamos la muestra no representativa 


In [ ]:
data = df.query('organismo == "Ecobici"')
data = pd.DataFrame(((data.isnull().sum() / len(data))*100))
data = pd.DataFrame(data)
data = data.reset_index()
data.columns = ['columna','porcentaje']


In [ ]:
fig2 = px.bar(data, x='columna', y='porcentaje')

In [ ]:
fig2

In [ ]:
def vn(orga,titulo):
    #data = df.query('organismo' == orga)
    data = df[df.organismo == orga]
    data = pd.DataFrame(((data.isnull().sum() / len(data))*100))
    data = pd.DataFrame(data)
    data = data.reset_index()
    data.columns = ['columna','porcentaje']
    return px.bar(data, x='columna', y='porcentaje',title = titulo)
   

In [ ]:
vn("Metrobús","qw")

In [ ]:

fig

Instrucciones
Responde a las siguientes preguntas: 
¿Cómo se distribuyen las columnas de afluencia_tarjeta, afluencia_boleto y afluencia_total_preliminar? ¿En dónde está el promedio? ¿Se concentran más en los valores altos o bajos? Crea un histograma con 40 bins para cada columna, uno encima del otro (3 renglones, 1 columna). 


In [ ]:
fig = make_subplots(rows=3, cols=1,
                    shared_xaxes=False,
                    vertical_spacing=.1,
                    subplot_titles=("Afluencia total preliminar","Afluencia boleto", "Afluencia tarjeta"),
                   )
                
fig.add_trace(
    go.Histogram(x=df["afluencia_tarjeta"],nbinsx =40, name="Afluencia tarjeta",),
              row=3, col=1)

fig.add_trace(
    go.Histogram(x=df["afluencia_boleto"],nbinsx =40, name="Afluencia boleto"),
              row=2, col=1)

fig.add_trace(
    go.Histogram(x=df["afluencia_total_preliminar"],nbinsx =40, name="Afluencia total preliminar" ),
              row=1, col=1)

fig.update_layout(height=900, width=1000,
                  title_text="Distribución de las columnas de afluencia")


fig

In [ ]:
plott = go.Figure()
plott.add_trace(go.Histogram(x=df["afluencia_tarjeta"], name="Afluencia tarjeta",))
plott.add_trace(go.Histogram(x=df["afluencia_boleto"], name="Afluencia boleto"))
plott.add_trace(go.Histogram(x=df["afluencia_total_preliminar"], name="Afluencia total preliminar" ))


Instrucciones
Responde a las siguientes preguntas: 
Para continuar con el análisis de estas columnas, crea un boxplot de dichas columnas. 


In [ ]:
plot = go.Figure()
plot.add_trace(go.Box(x=df["afluencia_tarjeta"], name="afluencia tarjeta"))
plot.add_trace(go.Box(x=df["afluencia_boleto"],name="afluencia boleto"))
plot.add_trace(go.Box(x=df["afluencia_total_preliminar"],name="afluencia total preliminar"))
plot.update_xaxes(type="log")


In [ ]:
plot = go.Figure()
plot.add_trace(go.Box(x=df["afluencia_tarjeta"], name="afluencia tarjeta", boxpoints="outliers"))

plot.update_xaxes(type="log")

In [ ]:
plot = go.Figure()
plot.add_trace(go.Box(x=df["afluencia_boleto"],name="afluencia boleto"))
plot.update_xaxes(type="log")

In [ ]:
plot = go.Figure()

plot.add_trace(go.Box(x=df["afluencia_total_preliminar"],name="afluencia total preliminar"))
plot.update_xaxes(type="log")

In [ ]:
fig3 = px.density_heatmap(df, x="dia", y="organismo",z="afluencia_total_preliminar", marginal_y="histogram", )
fig3.show()

In [ ]:
fig4 = px.density_heatmap(df, x="organismo", y="linea_servicio",z="afluencia_total_preliminar", )
fig4.show()

In [ ]:
lm = pd.DataFrame(df.groupby(["linea_servicio"])["afluencia_total_preliminar"].sum().sort_values(ascending = False))
lm =lm.reset_index()

In [ ]:
lm

linea_servicio  \
0                                                  L1   
1                                                  L2   
2                                                  L3   
3                                                  LB   
4                                                  L5   
5                                                  L8   
6                                                  L7   
7                                                  LA   
8                                                 L12   
9                                                  L9   
10                                                 L6   
11                                          Ordinario   
12                                                 L4   
13                                            Expreso   
14                                     L1 Eje Central   
15                                Xochimilco-Tasqueña   
16                                             Ecobús   
17                                             Atenea   
18                               L4 Pto Aéreo-Rosario   
19                               L7 Lomas Estrella-CU   
20                                SEFI Metro Línea 12   
21                                       L2 Eje 2 sur   
22                                       L3 Eje 7 Sur   
23                               L5 Sn Felipe-Hidalgo   
24                             L6 Rosario-Chapultepec   
25                                         SEFI Metro   
26                       L9 Iztacalco-Villa de Cortés   
27                            LT1 (Atlalilco-Tláhuac)   
28                     L1 Campos Revolución-Tlalpexco   
29                                    Expreso Directo   
30                                           Nocturno   
31                                         Eco Atenea   
32                           Temporal Zaragoza-Tacuba   
33  Temporal Peñón-Panteón San Isidro-Metro Camarones   
34                               LT2 (Coyuya-Tláhuac)   
35                                    L8 Circuito IPN   
36                        Temporal Tlalpan-Xochimilco   
37                          Temporal Viveros-Tetelpan   
38                      Servicio apoyo Metro Línea 12   

    afluencia_total_preliminar  
0                  282218802.0  
1                  181470897.0  
2                  168816310.0  
3                  101715310.0  
4                   94596366.0  
5                   90264901.0  
6                   89482578.0  
7                   83532311.0  
8                   78835682.0  
9                   78760298.0  
10                  71889214.0  
11                  45580256.0  
12                  35249988.0  
13                  24647012.0  
14                  22888645.0  
15                  12306738.0  
16                   9181715.0  
17                   4765065.0  
18                   4632121.0  
19                   4558353.0  
20                   3392169.0  
21                   3123695.0  
22                   2997578.0  
23                   2963173.0  
24                   2382278.0  
25                   2159852.0  
26                    624598.0  
27                    514385.0  
28                    385048.0  
29                    368123.0  
30                    245724.0  
31                    186653.0  
32                    166094.0  
33                    135717.0  
34                     87252.0  
35                     75587.0  
36                     52560.0  
37                     49887.0  
38                         0.0

In [ ]:
lm= lm.assign(porcentaje=lm[["afluencia_total_preliminar"]]/lm[["afluencia_total_preliminar"]].sum()*100)

In [ ]:
lm

linea_servicio  \
0                                                  L1   
1                                                  L2   
2                                                  L3   
3                                                  LB   
4                                                  L5   
5                                                  L8   
6                                                  L7   
7                                                  LA   
8                                                 L12   
9                                                  L9   
10                                                 L6   
11                                          Ordinario   
12                                                 L4   
13                                            Expreso   
14                                     L1 Eje Central   
15                                Xochimilco-Tasqueña   
16                                             Ecobús   
17                                             Atenea   
18                               L4 Pto Aéreo-Rosario   
19                               L7 Lomas Estrella-CU   
20                                SEFI Metro Línea 12   
21                                       L2 Eje 2 sur   
22                                       L3 Eje 7 Sur   
23                               L5 Sn Felipe-Hidalgo   
24                             L6 Rosario-Chapultepec   
25                                         SEFI Metro   
26                       L9 Iztacalco-Villa de Cortés   
27                            LT1 (Atlalilco-Tláhuac)   
28                     L1 Campos Revolución-Tlalpexco   
29                                    Expreso Directo   
30                                           Nocturno   
31                                         Eco Atenea   
32                           Temporal Zaragoza-Tacuba   
33  Temporal Peñón-Panteón San Isidro-Metro Camarones   
34                               LT2 (Coyuya-Tláhuac)   
35                                    L8 Circuito IPN   
36                        Temporal Tlalpan-Xochimilco   
37                          Temporal Viveros-Tetelpan   
38                      Servicio apoyo Metro Línea 12   

    afluencia_total_preliminar  porcentaje  
0                  282218802.0   18.748306  
1                  181470897.0   12.055440  
2                  168816310.0   11.214773  
3                  101715310.0    6.757132  
4                   94596366.0    6.284208  
5                   90264901.0    5.996461  
6                   89482578.0    5.944490  
7                   83532311.0    5.549203  
8                   78835682.0    5.237197  
9                   78760298.0    5.232189  
10                  71889214.0    4.775731  
11                  45580256.0    3.027979  
12                  35249988.0    2.341721  
13                  24647012.0    1.637346  
14                  22888645.0    1.520534  
15                  12306738.0    0.817559  
16                   9181715.0    0.609958  
17                   4765065.0    0.316552  
18                   4632121.0    0.307720  
19                   4558353.0    0.302820  
20                   3392169.0    0.225348  
21                   3123695.0    0.207513  
22                   2997578.0    0.199135  
23                   2963173.0    0.196849  
24                   2382278.0    0.158259  
25                   2159852.0    0.143483  
26                    624598.0    0.041493  
27                    514385.0    0.034172  
28                    385048.0    0.025579  
29                    368123.0    0.024455  
30                    245724.0    0.016324  
31                    186653.0    0.012400  
32                    166094.0    0.011034  
33                    135717.0    0.009016  
34                     87252.0    0.005796  
35                     75587.0    0.005021  
36                     52560.0    0.003492  
37                     49887.0    0.003314  
38 

In [ ]:
lmn = pd.DataFrame(df.groupby(["organismo"])["afluencia_total_preliminar"].sum().sort_values(ascending = False))
lmn = lmn.reset_index()
lmn.afluencia_total_preliminar.sum()

1509975753.0

In [ ]:
1022922000/1509975753

0.6774426661935942

Instrucciones
Responde a las siguientes preguntas: 
¿Cuál es la línea de servicio más utilizada? ¿Qué porcentaje de todas las líneas constituye esta línea? ¿El cierre de la línea 1 del Metro afectará fuertemente a los usuarios?


La linea más utilizada es la línea 1 y el cierre de esta afectaría el 18% de la movilidad publica en la cdmx asumiendo que esta no afecta a otras lineas

Instrucciones :
Regresando al tratamiento de valores nulos, como un experimento, elimina todos los registros / renglones que tengan al menos un valor nulo en alguna columna. Si hacemos esto, ¿qué porcentaje de datos limpios tendríamos con respecto al original?
Hint: utiliza df.dropna(how = “any”)


In [ ]:
dfl = df.dropna(how= "any")


In [ ]:
dfl

id organismo       linea_servicio      dia      fecha  \
3950  3951       RTP            Ordinario  Domingo 2020-01-03   
3951  3952       RTP              Expreso  Domingo 2020-01-03   
3953  3954       RTP               Ecobús  Domingo 2020-01-03   
3954  3955       RTP             Nocturno  Domingo 2020-01-03   
3956  3957       RTP      Expreso Directo  Domingo 2020-01-03   
...    ...       ...                  ...      ...        ...   
7606  7607       RTP              Expreso   Jueves 2021-01-07   
7608  7609       RTP               Ecobús   Jueves 2021-01-07   
7609  7610       RTP             Nocturno   Jueves 2021-01-07   
7611  7612       RTP      Expreso Directo   Jueves 2021-01-07   
7612  7613       RTP  SEFI Metro Línea 12   Jueves 2021-01-07   

      afluencia_tarjeta  afluencia_boleto  afluencia_total_preliminar  
3950               51.0           85298.0                     85349.0  
3951             1537.0           42745.0                     44282.0  
3953              555.0           16231.0                     16786.0  
3954               59.0            7531.0                      7590.0  
3956                1.0               0.0                         1.0  
...                 ...               ...                         ...  
7606             5442.0           51905.0                     57347.0  
7608             2525.0           20658.0                     23183.0  
7609               43.0             642.0                       685.0  
7611              218.0            1490.0                      1708.0  
7612             2484.0           53686.0                     56170.0  

[2686 rows x 8 columns]

18714 

In [ ]:
2686/18714

0.14352890883830288

Al hacer eliminar todas las filas con datos nulos solo nos quedaría el 14.35% de todos los datos

Instrucciones :
¿Cuántos valores duplicados hay? Después, elimina todos los valores duplicados.
Hint: utiliza .duplicated() para revisar cuántos duplicados hay, y después, drop_duplicates() para eliminarlos
 



In [ ]:
dfd=df.drop_duplicates()

In [ ]:
dfd

id  organismo linea_servicio        dia      fecha  \
0          1    Ecobici            NaN    Domingo 2020-01-03   
1          2    Ecobici            NaN      Lunes 2020-02-03   
2          3    Ecobici            NaN     Martes 2020-03-03   
3          4    Ecobici            NaN  Miércoles 2020-04-03   
4          5    Ecobici            NaN     Jueves 2020-05-03   
...      ...        ...            ...        ...        ...   
18709  18710  Suburbano             L1      Lunes 2021-06-28   
18710  18711  Suburbano             L1     Martes 2021-06-29   
18711  18712  Suburbano             L1  Miércoles 2021-06-30   
18712  18713  Suburbano             L1     Jueves 2021-01-07   
18713  18714  Suburbano             L1    Viernes 2021-02-07   

       afluencia_tarjeta  afluencia_boleto  afluencia_total_preliminar  
0                    NaN               NaN                     11238.0  
1                    NaN               NaN                     29475.0  
2                    NaN               NaN                     31855.0  
3                    NaN               NaN                     31477.0  
4                    NaN               NaN                     31493.0  
...                  ...               ...                         ...  
18709                NaN               NaN                         NaN  
18710                NaN               NaN                         NaN  
18711                NaN               NaN                         NaN  
18712                NaN               NaN                         NaN  
18713                NaN               NaN                         NaN  

[18714 rows x 8 columns]

Al revisar los datos notamos que no hay datos duplicados en este dataFrame, por ende no hubo ningún dropeo de data

Instrucciones :
Agrupa por días, y suma la columna de afluencia_total_preliminar para todos los organismos y líneas de servicio. De esta forma tendrás un dataframe con dos columnas: la fecha y la afluencia_total_preliminar de todos los organismos. Grafica estos en una serie de tiempo. ¿Afectó el COVID-19 al uso del transporte público? ¿Sí o no? ¿Se puede notar algún incremento o decremento en el uso por la salida o introducción de las diferentes olas? Copia el código para generar esta gráfica aquí



In [ ]:
udf = pd.DataFrame(df.groupby(["fecha"])["afluencia_total_preliminar"].sum())
udf = udf.reset_index()

In [ ]:
udf

fecha  afluencia_total_preliminar
0   2020-01-03                   3527802.0
1   2020-01-04                   3108214.0
2   2020-01-05                   1487336.0
3   2020-01-06                   2200122.0
4   2020-01-07                   2872869.0
..         ...                         ...
484 2021-12-02                   3361359.0
485 2021-12-03                   3809198.0
486 2021-12-04                   3558032.0
487 2021-12-05                   3619481.0
488 2021-12-06                   3384753.0

[489 rows x 2 columns]

In [ ]:
fig5 = go.Figure([go.Scatter(x=udf['fecha'], y=udf['afluencia_total_preliminar'])])

In [ ]:
fig5

Según los datos el covid si afecto al uso del transporte publico en la ciudad de mexico

In [ ]:
df

id  organismo linea_servicio        dia      fecha  \
0          1    Ecobici            NaN    Domingo 2020-01-03   
1          2    Ecobici            NaN      Lunes 2020-02-03   
2          3    Ecobici            NaN     Martes 2020-03-03   
3          4    Ecobici            NaN  Miércoles 2020-04-03   
4          5    Ecobici            NaN     Jueves 2020-05-03   
...      ...        ...            ...        ...        ...   
18709  18710  Suburbano             L1      Lunes 2021-06-28   
18710  18711  Suburbano             L1     Martes 2021-06-29   
18711  18712  Suburbano             L1  Miércoles 2021-06-30   
18712  18713  Suburbano             L1     Jueves 2021-01-07   
18713  18714  Suburbano             L1    Viernes 2021-02-07   

       afluencia_tarjeta  afluencia_boleto  afluencia_total_preliminar  
0                    NaN               NaN                     11238.0  
1                    NaN               NaN                     29475.0  
2                    NaN               NaN                     31855.0  
3                    NaN               NaN                     31477.0  
4                    NaN               NaN                     31493.0  
...                  ...               ...                         ...  
18709                NaN               NaN                         NaN  
18710                NaN               NaN                         NaN  
18711                NaN               NaN                         NaN  
18712                NaN               NaN                         NaN  
18713                NaN               NaN                         NaN  

[18714 rows x 8 columns]